R/teal_lockfile.R

```diff
@@ -39,50 +39,23 @@ teal_lockfile <- function() {
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
-    }
-
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  shiny::onStop(function() file.remove(lockfile_path))
+  callr::r_bg(create_renv_lockfile, args = list(lockfile_path = lockfile_path))
```

Comment:
are you able to pass R options? This might be relevant for the lockfile creation

Reply:
For sure `.Renvrion` is shared between R and `callr` sessions. Need to double check what happens with `options` or `env` (environment variables).

Reply:
I was more referring to the options set during the session. Say, an `app.R` script that has `options()` calls followed by `teal::init()` (inside which we have callr call).

Reply:
Yes, need to double check `options()`. I think this is the most crucial, next to environmental variables

Reply:
Environmental Variables are propagated

```r
> library(callr)
> rx <- r_bg(function() print(Sys.getenv('RSTUDIO_USER_IDENTITY')))
> rx$read_all_output()
[1] "[1] \"kosinsm4\"\r\n"
```

Reply:
`renv` uses `config`, `paths` and `settings` S3 objects in its setup. Those are not passed to `callr` unless you specify to set function environment to the specific package

```r
library(callr)
rx <- r_bg(function() print(names(settings)))
rx$read_all_output()
[1] ""
rx <- r_bg(function() print(names(settings)), package = 'renv')
rx$read_all_output()
> [1] " [1] \"bioconductor.version\"      \"ignored.packages\"         \r\n [3] \"external.libraries\"        \"package.dependency.fields\"\r\n [5] \"ppm.enabled\"               \"ppm.ignored.urls\"         \r\n [7] \"r.version\"                 \"snapshot.type\"            \r\n [9] \"use.cache\"                 \"vcs.manage.ignores\"       \r\n[11] \"vcs.ignore.cellar\"         \"vcs.ignore.library\"       \r\n[13] \"vcs.ignore.local\"         \r\n"
```

Reply:
Sounds like a bigger problem but maybe there is a way to pass `teal` as a reference package, that would somehow load/use `renv::settings`, `renv::config` and `renv::paths`. However this is a potential blocker

Reply:
I think we need to pass `options()`, `renv::settings`, `renv::config` and `renv::paths` as parameters to `create_renv_lockfile`

Reply:
@pawelru made a good point with `options`, it looks like options are not passed from the current session (see code below). I think we don't need to pass all the options but only those needed by `renv` to correctly reproduce packages (this is why `repos` argument needs to be specified). This is why we don't need all options but only some:

1. app is hosted on the remote server where some options are set by `.Rprofile` file
2. app is hosted on the remote server where some envvars are set by `.Renviron` file 
3. app developer can set `options()` and `Sys.setenv()` (they should be included in the `qenv` as it is part of the preprocessing)
4. app is run with app the options and envvars combined.
5. app user runs the app and need to restore the default (system) environment. This is given by the `.Rprofile` and `.Renviron` which seems to be included in the `r_bg` by default.
6. If `renv.lock` includes sysvars and options then user is able to reproduce initial environment (before preprocessing)
7. `qenv` code includes necessary `options` and `envvars`  

We need to make sure that (5) and (6) are true, (7) is the responsibility of app developer.


```r
library(callr)
options(example_option = "hi there")
rx <- r_bg(function() {
  print(getOption("repos")) # this is from .Rprofile
  print(getOption("example_option")) # this is from current session
})
rx$read_all_output() |> cat()

#                                devel                              release 
# "https://pharmaverse.r-universe.dev"        "https://cloud.r-project.org" 
#                                CRAN 
#       "https://cloud.r-project.org" 
# 
# NULL
```

Reply:
> `options()` **do not** get propagated by default (maybe there is a way to manually propagate them
> 
> ```r
> library(callr)
> options(user = 'MARCIN')
> rx <- r_bg(function() print(getOption('user', default = 'NOUSER')))
> rx$read_all_output()
> > [1] "[1] \"NOUSER\"\r\n"
> ```

Are you sure you didn't have an option `user` by default but make with different value?

Reply:
Passing all options is easier then some options because you can set multiple options with one call.

```r
# getting options
session_options <- options()
# setting options
options(session_options)
```

```r
library(callr)

rx <- r_bg(function(the_options) {print(options(the_options))}, list(the_options = options()))
rx$read_all_output()
```

So setting/passing options is fairy easy. What is problematic is passing `renv::settings`, `renv:;config` and `renv::config` that are locked bindings. IMHO `renv` doesn't utilize any options.

Reply:
I think those 4 should be enough for `renv` to have the same setup as in the main session

<img width="564" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/bb60e9e8-5f59-4e89-80f2-66069e59bc0c">


However those S3 objects in `renv` package are locked

```r
> assign('settings', value = settings, envir = getNamespace('renv'))
Error in assign("settings", value = settings, envir = getNamespace("renv")) : 
  cannot change value of locked binding for 'settings'
```

Reply:
> IMHO renv doesn't utilize any options.

Uses but only `getOption("repos")`

<img width="588" alt="image" src="https://github.com/insightsengineering/teal/assets/6959016/bfb30de3-06cf-4347-9891-bfc10c8434da">

Reply:
> I think we don't need to pass all the options but only those needed by renv to correctly reproduce packages

> Uses but only getOption("repos")

I disagree with the above. We don't really know right now what options are used. `renv` might call other packages that utilises its own set of options. I don't want to go through the whole functionality to identify list of options used. Moreover, this list might evolve in time so this introducing significant overhead on the maintenance. Please choose simple and low-maintenance solutions if possible.

2. Have you experimented with `package` argument of `callr::r_bg`. What would happen if you set it to `"renv"`?

Reply:
@pawelru fair point. So, passing all options and system vars to the `callr::r_bg` is needed.

Reply:
```
> assign('settings', value = settings, envir = getNamespace('renv')) Error in assign("settings", value = settings, envir = getNamespace("renv")) : cannot change value of locked binding for 'settings'
```

Can anyone explain me why to set `renv` settings? `renv` settings are changes through `options(renv.settings)` so passing `options` should be enough. Or not?

Reply:
- passing options is doable ✅ https://github.com/insightsengineering/teal/pull/1255#discussion_r1661366619
- system environmental variables are passed by default ✅ https://github.com/insightsengineering/teal/pull/1255#discussion_r1661140720

What is left to be passed are `renv::setting`, `renv::config` and `renv::paths`.

> Have you experimented with package argument of callr::r_bg. What would happen if you set it to "renv"?

Yes, I tried in here

<img width="544" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/97240257-5e09-465a-bb31-e3ee7818baaf">

we just need to check if this is the same `renv::setting` as in the main session or is this new `renv::setting` in the new session that is clean.

Reply:
I don't think there is anything like `options(renv.settings)`. `renv` uses it's own 3 objects to track its own options

Reply:
If we pass `package = 'renv'` then `renv::settings` and all objects from `renv` get propagated with their state from the main session

```r

library(callr)
settings$bioconductor.version('3.2222')
settings$bioconductor.version()
> [1] "3.2222"

rx <- r_bg(function() {print(settings$bioconductor.version())}, package = 'renv')
rx$read_all_output()
> [1] "[1] \"3.2222\"\r\n"
```

Reply:
> I don't think there is anything like `options(renv.settings)`. `renv` uses it's own 3 objects to track its own options

`?renv::settings`
<img width="596" alt="image" src="https://github.com/insightsengineering/teal/assets/6959016/e2c4a652-1adb-4930-a5d5-4ef35b196a50">

Reply:
Yes, but those `options` are empty

```r
> settings$snapshot.type()
[1] "explicit"
> options('renv.settings.snapshot.type')
$renv.settings.snapshot.type
NULL
````

```r
> settings$snapshot.type("explicit")
> settings$snapshot.type()
[1] "explicit"
> options('renv.settings.snapshot.type')
$renv.settings.snapshot.type
NULL
```

So I assume something is off with this `options` part. The description and examples states to use objects (`settings`, `config`, `paths`) directly

<img width="512" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/57584ea5-a27a-41dc-93e2-cc811f0f735f">

<img width="415" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/1f8916fc-8a5f-46a0-906d-473495b731cf">

Reply:
If you set them, renv will get updated

```r
> options(renv.settings.snapshot.type = 'marcin')
> settings$snapshot.type()
[1] "marcin"
```

but at the beginning of the session, everything in `options(renv.settings*)` is NULL.

If someone changes options via `settings$function()` those are not reflected in `options()`. It's the other way around: if something is set via `options()` then it is reflected in `settings$function()`. So for the case where users set their specification via `settings$function()`, we need to pass `settings` object.

Reply:
In the end we figured out that we can pass
- `options()` as a parameter to `create_renv_lockfile`
- `renv` namespace to `create_renv_lockfile` environment with `r_bg` parameter (`package  = 'renv'`)
- environmental variables from main session by default to the `callr` process with `r_bg` parameter (`env = NULL`)

```r
callr::r_bg(
      func = create_renv_lockfile,
      args = list(
        lockfile_path = lockfile_path,
        opts = options()
      ),
      # prints results to the console
      stdout = "",
      # renv setup is orchestrated by its special S3 objects: renv::settings, renv::config and renv::paths
      #     `package` parameter include `renv` namespace inside the environment of `func = create_renv_lockfile`
      package = "renv" ,
      # callr process will use environmental variables from the main session (parent process)
      env = NULL
    )
```